### PR TITLE
Rework authentication so that it's not Cognito-specific anymore

### PR DESCRIPTION
**Pull request by grzegorzpro**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,20 +6,28 @@ Created and Open Sourced thanks to [LocalBini.com](http://Localbini.com) by [pdy
...
-(simply paste from https://cognito-idp.{your aws region}.amazonaws.com/{your pool id}/.well-known/jwks.json) \
-`COGNITO_ALLOWED_CLIENTS` Allowed Cognito clients id's. String in format "id1,id2,id3". Used for verifying IdToken signature.\
-`PRINT_TRACEBACK` More verbose errors. Defaults False (set as "0" or "1").\
-`LOGGING_LEVEL` For logging level. Defaults INFO.\
-`EXPIRATION_KEY` for defining the expiration key used in authorization default "exp" according to JWT standard.
-`ALLOWED_ISS` for defining the allowed issuer of JWT - Security feature.
+## Configuration
+
+Lambdalizator can be configured using the following environment variables: 
+- `INTERNAL_AUTH_JWK` - a JSON Web Key that can be used for encoding and decoding internal JWTs.
```
Comment (songofroland) [inline README.md:12]: Maybe point out what internal JWT would/should be?

File: README.md [hunk 2]
```diff
@@ -6,20 +6,28 @@ Created and Open Sourced thanks to [LocalBini.com](http://Localbini.com) by [pdy
...
+  auth token will have the prefix removed. Defaults to False (set as "0" or "1").
+- `PRINT_TRACEBACK` - more verbose errors. Defaults to False (set as "0" or "1").
+- `LOGGING_LEVEL` - log level used in the application. Defaults to INFO.
+- `EXPIRATION_KEY` - for defining the expiration key used in authorization default "exp" according to JWT standard.
 
 ---
 **NOTE**
 
-Setting `COGNITO_PUBLIC_KEYS` or `COGNITO_ALLOWED_CLIENTS` implicitly enables Cognito authentication.
+Setting `ALLOWED_PUBLIC_KEYS` or `ALLOWED_AUDIENCES` implicitly enables Cognito authentication.
```
Comment (songofroland) [inline README.md:30]: Outdated note

File: lbz/authentication.py [hunk 3]
```diff
@@ -1,22 +1,26 @@
...
 from lbz.misc import get_logger
 
 logger = get_logger(__file__)
 
 STANDARD_CLAIMS = ("sub", "aud", "auth_time", "iss", "exp", "iat", "token_use")
 
+REMOVE_PREFIXES = os.environ.get('AUTH_REMOVE_PREFIXES') == '1'
+
+
+def remove_prefix(text: str):
```
Comment (songofroland) [inline lbz/authentication.py:14]: Please add returned type: `-> type`

File: README.md [hunk 4]
```diff
@@ -6,20 +6,28 @@ Created and Open Sourced thanks to [LocalBini.com](http://Localbini.com) by [pdy
...
+- `INTERNAL_AUTH_JWK` - a JSON Web Key that can be used for encoding and decoding internal JWTs.
+  It needs to be a JSON object that can be generated for example using [mkjwk](https://mkjwk.org/).
+- `ALLOWED_PUBLIC_KEYS` - a list of public keys that can be used for decoding auth tokens send in the
+  `Authentication` header. If you are using Cognito, you can use public keys from:
+  https://cognito-idp.{your aws region}.amazonaws.com/{your pool id}/.well-known/jwks.json.
+- `ALLOWED_AUDIENCES` - a list of audiences that will be used for verifying the JWTs send in the
+  `Authentication` header. It should be a comma-separated list of strings, e.g. `aud1,aud2`.
+  If not set, any audience will be considered valid.
+- `ALLOWED_ISS` - allowed issuer of JWT - Security feature. If not set, issuer will not be checked.
+- `AUTH_REMOVE_PREFIXES` - if enabled, all fields starting with a prefix (like `cognito:`) in the
```
Comment (songofroland) [inline README.md:21]: I think we should point out it's relevant for parsing JWT to an User instance - ultimately parsing raw JWT is not our goal - user instance does

File: lbz/authz.py [hunk 5]
```diff
@@ -154,18 +158,17 @@ def get_restrictions(self) -> dict:
     @staticmethod
     def sign_authz(authz_data: dict) -> str:
         """Generates JWT token"""
-        return jwt.encode(authz_data, CLIENT_SECRET, algorithm="HS512")
+        return jwt.encode(authz_data, INTERNAL_AUTH_JWK, algorithm="RS256", headers={"kid": INTERNAL_AUTH_JWK["kid"]})
 
     @staticmethod
     def decode_authz(token: str) -> dict:
```
Comment (songofroland) [inline lbz/authz.py:164]: Maybe this method is not needed now ?

File: lbz/jwt_utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+def get_matching_jwk(auth_jwt_token: str) -> dict:
+    try:
+        kid_from_jwt_header = jwt.get_unverified_header(auth_jwt_token)["kid"]
+        for key in PUBLIC_KEYS:
+            if key["kid"] == kid_from_jwt_header:
+                return key
+
+        logger.error(f"Required key not found in configuration.")
+        raise Unauthorized
+    except (JWTError, KeyError):
```
Comment (songofroland) [inline lbz/jwt_utils.py:28]: Probably there won't be JWTError now

File: lbz/jwt_utils.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    if not PUBLIC_KEYS:
+        logger.error('Invalid configuration - no keys in the ALLOWED_PUBLIC_KEYS env variable')
+        raise Unauthorized
+
+    jwk = get_matching_jwk(auth_jwt_token)
+    for aud in ALLOWED_AUDIENCES or [None]:
+        try:
+            return jwt.decode(auth_jwt_token, jwk, algorithms="RS256", audience=aud)
+        except JWTClaimsError:
+            pass
```
Comment (songofroland) [inline lbz/jwt_utils.py:42]: Intentional?

**Review by songofroland — APPROVED** [review]
Needs minor cosmetic changes

File: README.md (hunk 1 above)
Comment (grzegorzpro) [reply README.md:12]: I could add something, but to be honest I think this projects need a lot more documentation to explain the whole concept. Would it be ok to leave it for later? I'll have more time to write something in two weeks :)

File: README.md (hunk 2 above)
Comment (grzegorzpro) [reply README.md:30]: Actually it still works that way, check line 71 in resource.py

File: README.md (hunk 4 above)
Comment (grzegorzpro) [reply README.md:21]: Same situation, I'd do an overhaul of the docs separately

File: lbz/jwt_utils.py (hunk 7 above)
Comment (grzegorzpro) [reply lbz/jwt_utils.py:42]: Yes, this is the case when the audience is invalid, so we check with another audience

File: lbz/jwt_utils.py (hunk 6 above)
Comment (grzegorzpro) [reply lbz/jwt_utils.py:28]: Looks like `get_unverified_header` can still throw this error if the token is invalid

**Review by songofroland — APPROVED** [review]

**Pull request merged**